{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Gene_Expression_Dimensionality_Reduction.ipynb","provenance":[],"authorship_tag":"ABX9TyMk7s6X8miqBol8LaGfvy+a"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Gene Expressions RNAseq\n","\n","Visualizing high-dimensional datasets using Dimensionality Reduction techniques"],"metadata":{"id":"huLTw4L9BctL"}},{"cell_type":"markdown","source":["Use genomic [gene expression dataset](https://drive.google.com/drive/folders/1cFaJ0Oh-8-wv7xmUoC9xvdJ5NyujTfC8?usp=sharing) (6063 samples * 14460 identifiers). This dataset is already prepared for you by removing non-active genes which reduced its size from 60499 to 14460 features. If you are a domain expert, you can clean the dataset further. Otherwise, you can use the one provided.  \n","The original dataset was taken from [here](https://xenabrowser.net/datapages/?dataset=tcga_RSEM_gene_tpm&host=https%3A%2F%2Ftoil.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443).  \n","\n","- Visualize your dataset in a scatter plot by projecting it via t-SNE and/or UMAP, PCA or another dimensionality reduction technique into a 2-dimensional space.\n","- This article on [Visualizing high-dimensional datasets using PCA and t-SNE in Python](https://towardsdatascience.com/visualising-high-dimensional-datasets-using-pca-and-t-sne-in-python-8ef87e7915b) is a good starting point. Read carefully the article on [How to Use t-SNE Effectively](https://distill.pub/2016/misread-tsne/), so that you don't get tricked by your plots!\n"," \n","- __Perform a principal component analysis (PCA).__\n","    - How many principal components represent 95% of the total variance?\n","    - Apply clustering after PCA. \n","    - Analyze and interpret the results by creating visuals. \n","    - Color the individual points from the 2-D scatter plot according to the cluster they belong to.\n","\n","\n","- Redo the clustering by using [UMAP (Uniform Manifold Approximation and Projection)](https://umap-learn.readthedocs.io/en/latest/index.html) for dimensionality reduction.\n","\n","- Compare the cluster you found with the [primary_disease](https://drive.google.com/file/d/1-JrZ42o6kjK_jthPXYMdR-FY7CbqP-1m/view?usp=sharing) label from clinicalMatrix dataset. Hint: use beautiful [Sankey Diagrams](https://plotly.com/python/sankey-diagram/)\n","- Using the plots (scatterplot and Sankey), compare the effect of different dimensionality reduction methods on the clustering."],"metadata":{"id":"262wxYpqBr8q"}},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"id":"8v-kFNa8C4vY"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# auto reload packages and modules when they are modified\n","%load_ext autoreload\n","%autoreload 2\n","# %load_ext lab_black"],"metadata":{"id":"pJmChujHC89q"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":[""],"metadata":{"id":"xXbeG7TxCh3-"}},{"cell_type":"markdown","source":[""],"metadata":{"id":"Tm-SYPTyCh1N"}},{"cell_type":"markdown","source":[""],"metadata":{"id":"G6x0h19bChyv"}},{"cell_type":"markdown","source":[""],"metadata":{"id":"B7jYTKWNChv4"}},{"cell_type":"markdown","source":[""],"metadata":{"id":"F7lsC26QChrA"}},{"cell_type":"markdown","source":[""],"metadata":{"id":"9_2FB09UChoY"}},{"cell_type":"markdown","source":[""],"metadata":{"id":"DEnRM5WZChl0"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"pEOWqJ2vBRBx"},"outputs":[],"source":["# GENERAL\n","import pandas as pd\n","import numpy as np\n","import matplotlib.pyplot as plt\n","# OTHER\n","import warnings\n","warnings.filterwarnings(\"ignore\")\n","from sklearn.preprocessing import StandardScaler\n","from sklearn.preprocessing import LabelEncoder\n","# PIPELINE AND MODELS\n","from sklearn.pipeline import Pipeline\n","from sklearn.decomposition import PCA\n","from sklearn.cluster import KMeans\n","# VISUALIZATION\n","from sklearn.manifold import TSNE  \n","from umap import UMAP\n","from plotly import graph_objs as go\n","plt.style.use(\"ggplot\")\n","plt.rcParams[\"figure.figsize\"] = (12, 8)"]},{"cell_type":"code","source":["def rows_and_columns_counter (df):\n","    print('Imported DataFrame has the following shape:', df.shape)\n","    print('\\tso **', len(df.index),'** rows')\n","    print('\\tso **', len(df.columns),'** columns')    "],"metadata":{"id":"QLG23-oLCc76"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# FUNCTION --> Ploting distributions of sample columns\n","def plot_distrib(df, ncols, nrows, figsize):\n","    vars_ = df.columns\n","    c = 0\n","    plt.figure(figsize=figsize)\n","    for v in vars_:\n","        ax = plt.subplot(nrows, ncols, c + 1)\n","        _, nbins, _ = plt.hist(df[v], 20, color=\"g\", alpha=0.6, label=v)\n","        plt.xlabel(v)\n","        plt.legend(loc=\"best\")\n","        c = c + 1\n","    plt.show()"],"metadata":{"id":"AmqjBGHsCc59"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Google colab loading file\n","from google.colab import files\n","files.upload()"],"metadata":{"id":"88heeb7LCc29"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["gene_express = pd.read_csv('tcga_RSEM_gene_tpm_sample_dataset.csv',index_col=0)"],"metadata":{"id":"P6CiN9rhD9NX"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# path from local notebook\n","# import repo_path as rp\n","# path_repo = rp.paths(path)\n","# gene_express = pd.read_csv('{}Data/tcga_RSEM_gene_tpm_sample_dataset.csv'.format(path_repo),index_col=0)"],"metadata":{"id":"eRVzyzvlCc0n"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[""],"metadata":{"id":"6HlezaekCcx0"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"6Mby62beBWhh"},"source":["# Data Exploration"]},{"cell_type":"markdown","metadata":{"id":"-vWGoO9bBWhh"},"source":["<span style=\"color:green\">**Data Frame**"]},{"cell_type":"code","source":["#the data has more columns than rows. \n","#this indicates that you should look into dimensionality reduction methods\n","\n","\n","rows_and_columns_counter (gene_express)"],"metadata":{"id":"LHgTMIoQCcvn"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"qHpmyp8TBWhi"},"outputs":[],"source":["gene_express.head(3)"]},{"cell_type":"markdown","metadata":{"id":"-FAR3CQTBWhi"},"source":["<span style=\"color:green\">**DataFrame Distributions Visualization:** Plot distribution of sample columns\n","    \n","- As there are 14460 columns, we cannot look at every feature individually.\n","- Here we take a random sample of only 15 columns and look at their distributions. \n","- This is to get a better feel of what the data looks like. \n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Xo-zd3h5BWhi"},"outputs":[],"source":["data = gene_express.sample(n=15,axis='columns')\n","\n","plot_distrib(data, ncols=5, nrows=3, figsize=(70, 50))"]},{"cell_type":"markdown","metadata":{"id":"hOP6eWe1BWhj"},"source":["# Dimensionality Reduction for Visualisation"]},{"cell_type":"markdown","metadata":{"id":"8DMMUxQbBWhj"},"source":["## t-SNE\n","\n","- **Important note:** bear in mind that t-SNE is mainly used for visualization purposes. So you shall not include it within the Pipeline"]},{"cell_type":"markdown","metadata":{"id":"_en-8oEMBWhj"},"source":["<span style=\"color:green\">**Scatter Visualization in 2-D:** Using t-distributed Stochastic Neighbor Embedding  **(t-SNE)**\n","    \n","- (t-SNE) is a non-linear technique for dimensionality reduction that is particularly well suited for the visualization of high-dimensional datasets\n","- https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html\n","\n","TSNE Parameters:\n","- n_components = Dimension of the embedded space. We are projecting it onto 2 dimensions, so we can plot it in a scatter plot\n","\n","- perplexity = number of nearest neighbors that is used in other manifold learning algorithms. Consider selecting a value between 5 and 50. As the perplexity value increases, we generally observe a tendency towards clearer shapes in the visualization\n","\n","- n_iter = Maximum number of iterations for the optimization. Should be at least 250\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"xrbjB-nsBWhk"},"outputs":[],"source":["# t-SNE Definition: \n","tsne = TSNE(n_components=2, verbose=1, perplexity=50, n_iter=1000)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"u64e4sEDBWhk"},"outputs":[],"source":["#we fit and transform high dimensional dataset to the tsne object defined above\n","# Fit_Transform:\n","X_tsne = tsne.fit_transform(gene_express)\n","X_tsne"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"E9nDVEfOBWhk"},"outputs":[],"source":["#X_tsne is the embedding of the training data in low-dimensional space (2 dimensions)\n","\n","print('original data shape:', gene_express.shape) #(6063, 14460)\n","print('embedded data shape:', X_tsne.shape) #(6063, 2)"]},{"cell_type":"markdown","metadata":{"id":"RSAFUg79BWhk"},"source":["__t-SNE Visualization:__\n","- t-SNE maps the multi-dimensional data to a lower dimensional space and attempts to find patterns in the data by identifying observed clusters based on similarity of data points with multiple features.\n","- Now the data has been reduced to two dimensions, we can plot and visualize it. \n","\n","- This was not possible with the original 14460 columns\n","\n","- However, after this process, the input features are no longer identifiable, and you cannot make any inference based only on the output of t-SNE. Hence it is mainly a data exploration and visualization technique.\n","\n","\n","\n","__t-SNE compared to other techniques:__\n","- t-SNE does not scale well (PCA is much quicker) \n","- t-SNE is for visualization purposes only. Clustering on t-SNE is a bad idea because it doesn't preserve the global data structure. Within cluster distances are  meaninful, but cluster distances between clusters are not guaranteed\n","- We will only do clustering with PCA\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"t2t7zHBLBWhl"},"outputs":[],"source":["# Plot Result of t-SNE:\n","\n","fig = plt.figure(figsize =(15, 15))\n","\n","# plot original roll\n","# ax = fig.add_subplot(211, projection='3d')\n","# ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=color, cmap=plt.cm.Spectral)\n","# ax.set_title(\"Original data\")\n","\n","# plot projected roll\n","ax = fig.add_subplot()\n","ax.scatter(X_tsne[:, 0], X_tsne[:, 1], cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('t-SNE - Projected data')\n","plt.show()"]},{"cell_type":"markdown","metadata":{"id":"MGN4QsBuBWhl"},"source":["## UMAP"]},{"cell_type":"markdown","metadata":{"id":"ro3RzHUgBWhl"},"source":["<span style=\"color:green\">**Scatter Visualization in 2-D:** Using Uniform Manifold Approximation and Projection **(UMAP)**\n","    \n","    \n","- The biggest advantage of UMAP over t-SNE is the more optimal balance between local and global structure and the computational efficiency\n","\n","- https://umap-learn.readthedocs.io/en/latest/\n","\n","\n","\n","UMAP Parameters:\n","- n_neighbors = parameter to control how UMAP balances local versus global structure in the data. It does this by constraining the size of the local neighborhood UMAP will look at when attempting to learn the manifold structure of the data. \n","\n","    This means that low values of n_neighbors will force UMAP to concentrate on very local structure (potentially to the detriment of the big picture), while large values will push UMAP to look at larger neighborhoods of each point when estimating the manifold structure of the data, losing fine detail structure for the sake of getting the broader of the data.\n","\n"," \n","- n_components = the dimensionality of the reduced dimension space we will be embedding the data into\n","\n","\n","- min_dist = parameter controls how tightly UMAP is allowed to pack points together. It, quite literally, provides the minimum distance apart that points are allowed to be in the low dimensional representation. \n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"QU-PpvuMBWhl"},"outputs":[],"source":["# UMAP Definition:\n","umap_embeddings = UMAP(n_neighbors=30, n_components=2,  min_dist= 0.3)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZhXn9DIqBWhl"},"outputs":[],"source":["# Fit_Transform:\n","X_umap = umap_embeddings.fit_transform(gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"SSNeZlmIBWhm"},"outputs":[],"source":["# Plot Result of UMAP:\n","\n","\n","fig = plt.figure(figsize =(15, 15))\n","\n","\n","# plot oroginal roll\n","# ax = fig.add_subplot(211, projection='3d')\n","# ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=color, cmap=plt.cm.Spectral)\n","# ax.set_title(\"Original data\")\n","\n","# plot projected roll\n","ax = fig.add_subplot()\n","ax.scatter(X_umap[:, 0], X_umap[:, 1],cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('UMAP - Projected data')\n","plt.show()\n","\n","\n","#Think about how different the t-SNE and UMAP plots look, given that they are they same data!"]},{"cell_type":"markdown","metadata":{"id":"lraowGZ6BWhm"},"source":["# Dimensionality Reduction for Clustering"]},{"cell_type":"markdown","metadata":{"id":"UnppBkcTBWhm"},"source":["## Principal Component Analysis **(PCA)** \n","\n","<span style=\"color:green\"> Linear dimensionality reduction using Singular Value Decomposition of the data to project it to a lower dimensional space.\n","    \n","- The priority of PCA is to preserve large variances in the data using singular value decomposition. \n","- It tries to provide a minimum number of variables that keeps the maximum amount of variation or information about how the original data is distributed\n","\n","\n","- https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"lh8vDGnqBWhm"},"outputs":[],"source":["# PCA Definition:\n","pca = PCA()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"D4ahJGMDBWhm"},"outputs":[],"source":["# Pipeline Definition: 1º) Scaling --> 2º) Model\n","pipe_pca = Pipeline([(\"scl\", StandardScaler()), (\"pca\", pca)])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wit1_r0QBWhn"},"outputs":[],"source":["# Obtaining PCA Parameters:\n","pca.get_params()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"KgiZbkISBWhn"},"outputs":[],"source":["# Fit_transform:\n","gene_pca = pipe_pca.fit_transform(gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"z2ZH8Y3hBWhn"},"outputs":[],"source":["#The explained_variance_ratio_ attribute in sklearn returns the percentage of variance explained by each of the selected components.\n","#remember we are looking for the max variance between \n","\n","pipe_pca.named_steps[\"pca\"].explained_variance_ratio_\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"2CUoP3-PBWhn"},"outputs":[],"source":["#numpy cumsum returns the cumulative sum of the elements \n","#in this case, i\n","\n","evr = np.cumsum(pipe_pca.named_steps[\"pca\"].explained_variance_ratio_)\n","evr"]},{"cell_type":"markdown","metadata":{"id":"0EpeMT7lBWho"},"source":["<span style=\"color:green\">**PCA Visualization:** interceptor location"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"iVYRwlluBWho"},"outputs":[],"source":["# Finding out the interceptor that represents 95% of the total variance:\n","x_intercept = np.argmax(evr > 0.95) #evr is an array of the cumulative variances\n","print('The interceptor is located at component: ', x_intercept)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MMzfKpwTBWho"},"outputs":[],"source":["# Ploting the location of the interceptor:\n","\n","plt.plot(evr, \"-x\", color=\"lightblue\") #plotting the cumulative sum of variance, so we can where the variance increases are not as pronounced\n","plt.xlabel(\"Number of Components\")\n","plt.ylabel(\"Explained Variance Ratio\")\n","plt.axvline(x=x_intercept, color=\"red\") #vertical line to see the number of components that corresponds to the 95% total variance\n","plt.axhline(y=0.95, color=\"red\") #horizontal line for 95% of the total variance \n","None"]},{"cell_type":"markdown","metadata":{"id":"zYMv1-IPBWho"},"source":["<span style=\"color:green\">**PCA Scatter Visualization in 2-D:** PCA example components\n","    \n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"i9jMsu9QBWho"},"outputs":[],"source":["#After dimensionality reduction, we can use gene_pca plot the principal components \n","\n","plt.rcParams[\"figure.figsize\"] = (15, 15)\n","\n","fig = plt.figure()\n","\n","# # plot original roll\n","# ax = fig.add_subplot(311, projection='3d')\n","# ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=color, cmap=plt.cm.Spectral)\n","# ax.set_title(\"Original data\")\n","\n","# plot projected roll (components 1 and 2)\n","ax = fig.add_subplot(311)\n","ax.scatter(gene_pca[:, 0], gene_pca[:, 1],  cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('Projected data (components 1 and 2)')\n","\n","# plot projected roll (components 2 and 3)\n","ax = fig.add_subplot(312)\n","ax.scatter(gene_pca[:, 1], gene_pca[:, 2],cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('Projected data (components 2 and 3)')\n","plt.show()\n","\n","\n","# # plot projected roll (components 1 and 3)\n","# ax = fig.add_subplot(313)\n","# ax.scatter(gene_pca[:, 2], gene_pca[:, 3],cmap=plt.cm.Spectral)\n","# plt.axis('tight')\n","# plt.xticks([]), plt.yticks([])\n","# plt.title('Projected data (components 1 and 3)')\n","# plt.show()\n","\n","plt.show()"]},{"cell_type":"markdown","metadata":{"id":"6g3QdRnpBWhp"},"source":["# Clustering"]},{"cell_type":"markdown","metadata":{"id":"9iFj19u8BWhp"},"source":["<span style=\"color:green\">**K-Means after PCA:** select the ‘optimal’ number of clusters with the **Elbow Method (for a 95% of variance)**\n","    \n","- After using PCA for dimensionality reduction, we can use k-means clustering to find the number of clusters. \n","- Determining the ideal number of clusters for our k-means model can be done by measuring the sum of the squared distances to the nearest cluster center\n","- The k-means plot below indicates the percentage of variance explained, but in slightly different terms, as a function of the number of clusters.\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"8QErb5nEBWhp"},"outputs":[],"source":["r_seed = 23  # random seed to use during modeling for reproducibility \n","cluster_errors = []\n","\n","for i in range(1, 14):\n","    n_clusters = i\n","    pipe_pca_kmean = Pipeline(\n","        [\n","            (\"scl\", StandardScaler()), \n","            (\"pca\", PCA(0.95)), \n","            (\"cluster\", KMeans(n_clusters=n_clusters, random_state=r_seed, verbose=0, n_jobs=1))]\n","    )\n","\n","    pipe_pca_kmean.fit(gene_express)\n","    pipe_pca_kmean.predict(gene_express)\n","    cluster_errors.append(pipe_pca_kmean.named_steps[\"cluster\"].inertia_) "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"eOaFuPooBWhq"},"outputs":[],"source":["plt.clf()\n","plt.plot(cluster_errors, \"o-\")\n","plt.xlabel(\"n_clusters\")\n","plt.ylabel(\"sum sq distances from mean\")\n","plt.show()"]},{"cell_type":"markdown","metadata":{"id":"UyOxxsHOBWhq"},"source":["<span style=\"color:green\">**K-Means clustering with PCA:**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4qpkYE2sBWhr"},"outputs":[],"source":["# Pipeline Definition: 1º) Scaling --> 2º) Model --> 3º) Clusterization Method\n","\n","pipe_pca_kmean = Pipeline(\n","        [\n","            (\"scl\", StandardScaler()), \n","            (\"pca\", PCA(0.95)), \n","            (\"cluster\", KMeans(n_clusters=6, random_state=r_seed, verbose=0, n_jobs=1))]\n","    )\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hj1lB8cZBWhr"},"outputs":[],"source":["# Fit:\n","pipe_pca_kmean.fit(gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"OXjskgj2BWhr"},"outputs":[],"source":["gene_kmean_cluster = pd.DataFrame(\n","    data = pipe_pca_kmean.named_steps[\"cluster\"].labels_,\n","    index = gene_express.index,\n","    columns=['kmean_cluster']\n",")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"3Sd6eJ0wBWhs"},"outputs":[],"source":["gene_kmean_cluster['kmean_cluster'].value_counts()"]},{"cell_type":"markdown","metadata":{"id":"aErKpN3dBWhs"},"source":["<span style=\"color:green\">**Visualize cluster by t-SNE:**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GQAdCBiRBWhs"},"outputs":[],"source":["# tsne = TSNE(n_components=2, verbose=1, perplexity=50, n_iter=611)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"TOc-tZMhBWht"},"outputs":[],"source":["# X_tsne = tsne.fit_transform(gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"50EO5frgBWht"},"outputs":[],"source":["fig = plt.figure()\n","\n","# # plot original roll\n","# ax = fig.add_subplot(211, projection='3d')\n","# ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=color, cmap=plt.cm.Spectral)\n","# ax.set_title(\"Original data\")\n","\n","# plot projected roll\n","ax = fig.add_subplot()\n","ax.scatter(X_tsne[:, 0], X_tsne[:, 1], c= gene_kmean_cluster['kmean_cluster'],alpha=.5, cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('t-SNE Projected data with Kmean Cluster')\n","plt.show()\n","\n","\n","# plt.show()"]},{"cell_type":"markdown","metadata":{"id":"9jqTuvFXBWht"},"source":["<span style=\"color:green\">**Uniform Manifold Approximation and Projection (UMAP):** using elbow method"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bHo2Kph6BWhu"},"outputs":[],"source":["rows_and_columns_counter (gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LOp6KKtNBWhu"},"outputs":[],"source":["cluster_errors = []\n","for i in range(1, 14):\n","    n_clusters = i\n","    pipe_umap_kmean = Pipeline(\n","        [\n","            (\"scl\", StandardScaler()), \n","            (\"UMAP\", UMAP(n_neighbors=30, n_components=100,  min_dist= 0.3)), \n","            (\"cluster\", KMeans(n_clusters=n_clusters, random_state=r_seed, verbose=0, n_jobs=1))]\n","    )\n","    pipe_umap_kmean.fit(gene_express)\n","    pipe_umap_kmean.predict(gene_express)\n","    cluster_errors.append(pipe_umap_kmean.named_steps[\"cluster\"].inertia_) "]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ODtTqCHRBWhu"},"outputs":[],"source":["plt.clf()\n","plt.plot(cluster_errors, \"o-\")\n","plt.xlabel(\"n_clusters\")\n","plt.ylabel(\"sum sq distances from mean\")\n","plt.show()"]},{"cell_type":"markdown","metadata":{"id":"zaW_Gv7DBWhv"},"source":["<span style=\"color:green\">**K-mean after UMAP:**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"5dBe3BuFBWhv"},"outputs":[],"source":["pipe_umap_kmean = Pipeline(\n","        [\n","            (\"scl\", StandardScaler()), \n","            (\"UMAP\", UMAP(n_neighbors=30, n_components=100,  min_dist= 0.3)), \n","            (\"cluster\", KMeans(n_clusters=6, random_state=r_seed, verbose=0, n_jobs=1))]\n","    )"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"vgoNMBWbBWhv"},"outputs":[],"source":["pipe_umap_kmean.fit(gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DtGUD0pFBWhv"},"outputs":[],"source":["gene_umap_cluster = pd.DataFrame(\n","    data = pipe_umap_kmean.named_steps[\"cluster\"].labels_,\n","    index = gene_express.index,\n","    columns=['Cluster_id'])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"OvkyeqFkBWhv"},"outputs":[],"source":["gene_umap_cluster['Cluster_id'].value_counts()"]},{"cell_type":"markdown","metadata":{"id":"6cYyYYFRBWhv"},"source":["### Visualize cluster by t-SNE"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"WcfyDXelBWhv"},"outputs":[],"source":["# tsne = TSNE(n_components=2, verbose=1, perplexity=50, n_iter=611)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"OqLIQlb3BWhw"},"outputs":[],"source":["# X_tsne = tsne.fit_transform(gene_express)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"8Odc3cDsBWhw"},"outputs":[],"source":["fig = plt.figure()\n","\n","# # plot original roll\n","# ax = fig.add_subplot(211, projection='3d')\n","# ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=color, cmap=plt.cm.Spectral)\n","# ax.set_title(\"Original data\")\n","\n","# plot projected roll\n","ax = fig.add_subplot()\n","ax.scatter(X_tsne[:, 0], X_tsne[:, 1], c= gene_umap_cluster['Cluster_id'],alpha =0.5, cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('t-SNE Projected data with UMAP cluster')\n","plt.show()\n","\n","\n","# plt.show()"]},{"cell_type":"markdown","metadata":{"id":"ngLqZr2IBWhw"},"source":["### Compare with given labels"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"mXPjnRViBWhw"},"outputs":[],"source":["#these are the clusters that UMAP has given us\n","gene_umap_cluster.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9Y2AC03vBWhw"},"outputs":[],"source":["#phenotype_labels.csv has the true labels\n","phenotype = pd.read_csv('../data/phenotype_labels.csv',index_col=0)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"YagM7dw5BWhw"},"outputs":[],"source":["phenotype.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"6AWZAReHBWhw"},"outputs":[],"source":["le = LabelEncoder()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"iFAzEpO4BWhx"},"outputs":[],"source":["\n","phenotype['primary_disease_id'] = le.fit_transform(phenotype['_primary_disease'])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"42wQWEuQBWhx"},"outputs":[],"source":["phenotype.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"gKXSOoEtBWhx"},"outputs":[],"source":["phenotype['primary_disease_id'].value_counts()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"xxV7iXrlBWhx"},"outputs":[],"source":["le.classes_"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"c-0E_WaUBWhx"},"outputs":[],"source":["gene_umap_cluster['disease_id'] = phenotype['primary_disease_id']"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rW9A8O3xBWhx"},"outputs":[],"source":["gene_express.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"HoEbneKeBWhx"},"outputs":[],"source":["phenotype['primary_disease_id'].reindex(gene_express.index)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"t-wBT8cZBWhx"},"outputs":[],"source":["fig = plt.figure()\n","\n","# # plot original roll\n","# ax = fig.add_subplot(211, projection='3d')\n","# ax.scatter(X[:, 0], X[:, 1], X[:, 2], c=color, cmap=plt.cm.Spectral)\n","# ax.set_title(\"Original data\")\n","\n","# plot projected roll\n","ax = fig.add_subplot()\n","ax.scatter(X_tsne[:, 0], X_tsne[:, 1], c= phenotype['primary_disease_id'].reindex(gene_express.index),alpha =0.5, cmap=plt.cm.Spectral)\n","plt.axis('tight')\n","plt.xticks([]), plt.yticks([])\n","plt.title('t-SNE Projected data with real data')\n","plt.show()\n","\n","\n","# plt.show()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"5nqVyY53BWhy"},"outputs":[],"source":["gene_kmean_cluster['disease_id'] = phenotype['primary_disease_id']"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"IzJd-gaeBWhy"},"outputs":[],"source":["gene_kmean_cluster.head()"]},{"cell_type":"markdown","metadata":{"id":"31MyJ9moBWhy"},"source":["### Sankey Diagram Kmean"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"M-OzF5NfBWhy"},"outputs":[],"source":["sd = gene_kmean_cluster.groupby([\"kmean_cluster\", \"disease_id\"]).size().reset_index(name=\"value\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"n0-kphMwBWhy"},"outputs":[],"source":["len(sd['disease_id'].unique())"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"lD8hmIUaBWhy"},"outputs":[],"source":["sd['kmean_cluster'] = sd['kmean_cluster']+len(sd['disease_id'].unique())"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"WxUFqZ0BBWhy"},"outputs":[],"source":["sd.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"dU7FAjNoBWhz"},"outputs":[],"source":["sd_labels = list(le.classes_) +['KMEAN'+str(i) for i in range(len(sd['kmean_cluster'].unique()))]"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"FW88uEHZBWhz"},"outputs":[],"source":["sd_labels"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"JPhI4ULeBWhz"},"outputs":[],"source":["fig = go.Figure(\n","    data = [\n","    go.Sankey(\n","    node=dict(\n","        label= sd_labels\n","    ),\n","    link=dict(\n","        source=list(sd[\"disease_id\"]), target=list(sd[\"kmean_cluster\"]), value=list(sd[\"value\"])\n","    ),\n","\n","        )\n","    ]\n",")\n","       \n","\n","fig.update_layout(title_text=\" Kmeans Cluster Analysis\", font_size=15)\n","fig.show()"]},{"cell_type":"markdown","metadata":{"id":"EAvxBoqXBWhz"},"source":["### Sankey Diagram UMAP\n","<span style=\"color:green\"> A Sankey diagram is a flow diagram, in which the width of arrows is proportional to the flow quantity.\n","\n","\n","\n","https://plotly.com/python/sankey-diagram/"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"vQXxQAN3BWhz"},"outputs":[],"source":["sd_umap = gene_umap_cluster.groupby([\"Cluster_id\", \"disease_id\"]).size().reset_index(name=\"value\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZpRekWSQBWhz"},"outputs":[],"source":["len(sd_umap['disease_id'].unique())"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"b8pMo_kuBWhz"},"outputs":[],"source":["sd_umap['Cluster_id'] = sd_umap['Cluster_id']+len(sd_umap['disease_id'].unique())"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PjNcJldgBWh0"},"outputs":[],"source":["sd_umap.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yM-0mUbcBWh0"},"outputs":[],"source":["sd_umap_labels = list(le.classes_) +['UMAP'+str(i) for i in range(len(sd_umap['Cluster_id'].unique()))]"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"V-D7MO9xBWh0"},"outputs":[],"source":["sd_umap_labels"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Y5yg0N7wBWh0"},"outputs":[],"source":[" fig = go.Figure(\n","    data = [\n","    go.Sankey(\n","    node=dict(\n","        label= sd_umap_labels\n","    ),\n","    link=dict(\n","        source=list(sd_umap[\"disease_id\"]), target=list(sd_umap[\"Cluster_id\"]), value=list(sd_umap[\"value\"])\n","    ),\n","\n","        )\n","    ]\n",")\n","       \n","\n","fig.update_layout(title_text=\"UMAP Cluster Analysis\", font_size=15)\n","fig.show()"]},{"cell_type":"markdown","metadata":{"id":"k-QpoRhbBWh0"},"source":["--------------"]},{"cell_type":"code","source":[""],"metadata":{"id":"8XY_Gxf-CctC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[""],"metadata":{"id":"VMIE_gL4Ccqh"},"execution_count":null,"outputs":[]}]}